lib/makara/context.rb

```diff
@@ -1,83 +1,144 @@
-      # when they're asking whether they should be stuck to master.
-      def stick(context, config_id, ttl)
-        Makara::Cache.write(cache_key_for(context, config_id), '1', ttl)
+      def release(config_id)
+        current.release(config_id)
       end
 
-      def stuck?(context, config_id)
-        !!Makara::Cache.read(cache_key_for(context, config_id))
+      def release_all
+        current.release_all
       end
 
       protected
-
-      def previously_sticky
-        fetch(:makara_previously_sticky) { Hash.new }
-      end
-
-      def cache_key_for(context, config_id)
-        "makara::#{context}-#{config_id}"
+      def current
+        fetch(:makara_current_context) { new({}) }
       end
 
       def fetch(key)
-        get(key) || set(key,yield)
+        get(key) || set(key, yield)
       end
 
       if Thread.current.respond_to?(:thread_variable_get)
         def get(key)
           Thread.current.thread_variable_get(key)
         end
 
-        def set(key,value)
-          Thread.current.thread_variable_set(key,value)
+        def set(key, value)
+          Thread.current.thread_variable_set(key, value)
         end
       else
         def get(key)
           Thread.current[key]
         end
 
-        def set(key,value)
+        def set(key, value)
           Thread.current[key]=value
         end
       end
 
+      private
+
+      # Pairs of {config_id}:{timestamp}, separated by "|"
+      # config_id1:1518270031.3132212|config_id2:1518270030.313232 ..
+      def parse(cookie_string)
+        return {} if cookie_string.empty?
+
+        states = cookie_string.split("|")
+        states.each_with_object({}) do |state, data|
+          config_id, timestamp = state.split(":")
+          data[config_id] = timestamp.to_f if config_id && timestamp
+        end
+      end
+    end
+
+    attr_accessor :data
+
+    def initialize(data)
+      @data = data
+      @dirty = false
+    end
+
+    def stick(config_id, ttl)
+      data[config_id] = Time.now.to_f + ttl.to_f
+      @dirty = true
+    end
+
+    def stuck?(config_id)
+      data[config_id] && !expired?(data[config_id])
+    end
+
+    # Indicates whether there have been changes to the context that need
+    # to be persisted when the request finishes
+    def dirty?
+      @dirty
+    end
+
+    def release(config_id)
+      @dirty ||= !!data.delete(config_id)
+    end
+
+    def release_expired
+      previous_size = data.size
+      data.delete_if { |_, timestamp| expired?(timestamp) }
+      @dirty ||= previous_size != data.size
+    end
+
+    def release_all
+      if self.data.any?
+        self.data = {}
+        @dirty = true
+      end
+    end
+
+    def to_cookie_options
+      max_age = if data.any?
+        (data.values.max - Time.now.to_f).ceil + 1
+      else
+        0
+      end
+
+      value = data.collect { |config_id, ttl| "#{config_id}:#{ttl}" }.join('|')
+      { :max_age => max_age, :value => value }
```

Comment:
Good question! The current implementation uses `max-age` and I thought whether it should e changed to `expires`. From what I read, it seemed that `expires` was deprecated on HTTP version 1.1, replaced by `max-age`. I understood that this was simpler and better because it isn't affected by clock differences between the server and the client browser. However, older versions of IE don't support `max-age`, so `expires` is still used in many places to ensure browser support. 

I just realised that when setting both, `max-age` takes precedence, so perhaps I should just set both 😄

Reply:
Whatever you think as you've looked into it more.